applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nonlinear_strain_energy_response_function.cpp

```diff
@@ -0,0 +1,378 @@
+        int i = 0;
+        for (auto& node_i : rAdjointCondition.GetGeometry())
+        {
+            project(displacement, range(i, dimension)) = rAdjointCondition.GetGeometry()[node_i].FastGetSolutionStepValue(DISPLACEMENT);
+            project(displacement_previous_step , range(i , dimension))= rAdjointCondition.GetGeometry()[node_i].FastGetSolutionStepValue(DISPLACEMENT,1);   
+            i += 3;
+        }
+
+        mConditions[rAdjointCondition.Id()]->CalculateRightHandSide(RHS, process_info);
+        
+        //calculation of partial f_{ext}_i}{\partial u}
+        Matrix partial_derivative_matrix = ZeroMatrix(vec_size, vec_size);
+        int i_2 = 0;
+        for (auto& node_i : rAdjointCondition.GetGeometry())
+        {
+            Vector perturbed_RHS = Vector(0);
+
+            // Pertubation, gradient analysis and recovery of x
+            node_i.X() += delta; 
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_X) += delta;
+            mConditions[rAdjointCondition.Id()]->CalculateRightHandSide(perturbed_RHS, process_info);
+            row(partial_derivative_matrix, i_2) = (perturbed_RHS - RHS) / delta;
+            node_i.X() -= delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_X) -= delta;
+                
+            // Reset the pertubed vector
+            perturbed_RHS = Vector(0);
+
+            // Pertubation, gradient analysis and recovery of y
+            node_i.Y() += delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_Y) += delta;
+            mConditions[rAdjointCondition.Id()]->CalculateRightHandSide(perturbed_RHS, process_info);
+            row(partial_derivative_matrix, i_2 + 1) = (perturbed_RHS - RHS) / delta;
+            node_i.Y() -= delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_Y) -= delta;
+            
+            // Reset the pertubed vector
+            perturbed_RHS = Vector(0);
+
+            // Pertubation, gradient analysis and recovery of z
+            node_i.Z() += delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_Z) += delta;
+            mConditions[rAdjointCondition.Id()]->CalculateRightHandSide(perturbed_RHS, process_info);
+            row(partial_derivative_matrix, i_2 + 2) = (perturbed_RHS - RHS) / delta;
+            node_i.Z() -= delta;
+            node_i.FastGetSolutionStepValue(DISPLACEMENT_Z) -= delta;
+
+            i_2 += 3;
+        }
+
+        if(mExternalForceDisplacementDerivative.size1() == 0)
+            mExternalForceDisplacementDerivative = ZeroMatrix(vec_size, vec_size);
+        
+        // Summing up the partial derivative terms
+        rResponseGradient = 0.50 * (RHS + mConditionsRHS[rAdjointCondition.Id()] + 
+                            prod(displacement - displacement_previous_step, mExternalForceDisplacementDerivative + partial_derivative_matrix));
+        
+        // storing the RHS for each condition
+        mConditionsRHS[rAdjointCondition.Id()] = RHS;
```

Comment:
Id

Reply:
The alternative would be to create a new variable i guess